{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# W200 Introduction to Data Science Programming, UC Berkeley MIDS\n",
    "\n",
    "\n",
    "## Instructions\n",
    "The final exam is designed to evaluate your grasp of Python theory as well as Python coding.\n",
    "\n",
    "- This is an individual exam.\n",
    "- You have 48 hours to complete the exam, starting from the point at which you first access it.\n",
    "- You will be graded on the quality of your answers.  Use clear, persuasive arguments based on concepts we covered in class.\n",
    "- Please double-click the markdown cells where it says \"Your answer here\" to input answers (if you need more cells please make them markdown cells)\n",
    "- Use only Python standard libraries, matplotlib, seaborn, NumPy and Pandas for this exam\n",
    "- Upload your exam to gradescope and your github repo in the folder /SUBMISSIONS/final_exam/. **The code version uploaded to Part 2 will be your official version for grading purposes.**\n",
    "- There is no autograder for this notebook - your exam will be manually graded.\n",
    "- Please restart and run all cells before submitting so that all of your answers are shown. You do not need to upload the dataset into gradescope or github."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## YOUR NAME HERE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## 1: Short Answer Questions (25 pts - each question = 5 pts)\n",
    "\n",
    "a) The following class ```Cart``` and method ```add_to_cart``` are parts of a larger program used by a mobile phone company.  The method ```add_to_cart``` will work when an object of type MobileDevice or of type ServiceContract is passed to it. State whether the method ```add_to_cart``` is a demonstration of the following items (yes/no) and the reasoning (1-2 sentences): \n",
    "\n",
    "    1. Inheritance\n",
    "    2. Polymorphism\n",
    "    3. Object-Oriented programming\n",
    "    4. Top-down design\n",
    "    5. Functional programming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# Method:\n",
    "class Cart():\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.cart = []\n",
    "        self.total = 0\n",
    "    \n",
    "    def add_to_cart(self, item):\n",
    "        self.cart.append(item)\n",
    "        self.total += item.price"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "- a) Your answer here\n",
    "\n",
    "\n",
    "1. Inheritance: No, inheritance is when you have a parent and child class that inherits attributes and methods from the parent class, in this instance we have a single class.\n",
    "2. Polymorphism: Yes, there may be according ot the second sentenence that suggests this MobileDevice and ServiceContract object classes can be passed to the add_to_cart() method in this parent Cart() class.\n",
    "3. Object-Oriented programming: Yes, the objects attributes are called in the method with the keyword self, this is object oriented programming.\n",
    "4. Top-down design: No, we're not systematically breaking down a larger more complex problem into smaller more manageable sub problems.\n",
    "5. Functional programming: No, this is not functional programming because functional programming does not have a concept of class. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "b) Suppose you have a long list of digits (0-9) that you want to write to a file.  From a storage standpoint, would it be more efficient to use ASCII or UTF-8 as an encoding? What is the most efficient way to create an even smaller file to store the information?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "- b) From a storage standpoint when you have a long list of digits, theres no difference in storage size between ASCII or UTF-8. Huffman coding algorithm is a data compression algorithm that can create an even smaller file to store the information.  Unlike the either encoding that uses exactly eight bits to encode each character, a Huffman code uses a variable number of bits for each character."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "c) How would you sanity-check your data? Why is it important to sanity-check your data before you begin your analysis? What could happen if you don't?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "- c) You would need to check for datatype, or datatype mismatches, variations in how values are entered, missing values, and duplicate values or outliers in the data. Because wrong datasets can significantly impact the analysis and draw an incorrect solution or prediction. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "d) How do you determine which variables in your dataset you should check for issues prior to starting an analysis? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "- d) You would start by identifying which variables are most valuable to the analysis or data model using variable-importance measures."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "e1) Explain why the following code prints what it does."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'function'>\n"
     ]
    }
   ],
   "source": [
    "def g(): pass\n",
    "print(type(g))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "- e1) This is a function address, the code prints an empty function because the keyword def is used to define the function and the letter g is followed by parenthesis and semicolons."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "e2) Explain why the following code prints something different."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "def g(): pass\n",
    "print(type(g()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "- e2) This is a function call, it executes and generates the value pass which produces a NoneType class type. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## 2:  General Coding Questions (15 pts - each question 5 pts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "a) Using a dictionary comprehension, make a dictionary with the keys being the lowercase vowels (aeiou) and the values being the number of vowels in the example_text (case insensitive). Fill in a dictionary comprehension below so that we get this desired output.\n",
    "\n",
    "```\n",
    "example_text = 'Outlook is bright!'\n",
    "```\n",
    "\n",
    "expected output:\n",
    "```\n",
    "{'i': 2, 'o': 3, 'u': 1}\n",
    "\n",
    "or\n",
    "\n",
    "{'a': 0, 'e': 0, 'i': 2, 'o': 3, 'u': 1}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'a': 0, 'e': 1, 'i': 3, 'o': 3, 'u': 0}\n"
     ]
    }
   ],
   "source": [
    "example_text = \"I love coding in python!\"\n",
    "\n",
    "\n",
    "# 2a) Your code here\n",
    "example_dictionary = { letter : example_text.lower().count(letter) for letter in ['a', 'e', 'i', 'o', 'u'] }\n",
    "\n",
    "print(example_dictionary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "b) Below is a data frame of customers that have different cooling systems. Your data science team lead wants the column ```cooling_system``` to be labeled with the integer numbers 1-4 instead of the text as shown below:\n",
    "\n",
    "- 1 = Air Conditioning / Air Con / AC\n",
    "- 2 = Heat Pump / HP\n",
    "- 3 = Evaporative Cooler / EvapCooler / EC\n",
    "- 4 = Fan\n",
    "\n",
    "Make a new column called ```cooling_type``` that maps the text values to the new numeric values. Filter out / remove rows with the values that are not included in the mapping above. Print out/display this new data frame. Be sure to list any assumptions also!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cust_Number</th>\n",
       "      <th>Cooling_System</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>101</td>\n",
       "      <td>AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>102</td>\n",
       "      <td>Heat Pump</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>103</td>\n",
       "      <td>Air Con</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>104</td>\n",
       "      <td>Air Conditioning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>105</td>\n",
       "      <td>Fan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>106</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>107</td>\n",
       "      <td>EvapCooler</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>108</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>109</td>\n",
       "      <td>EC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>110</td>\n",
       "      <td>Evaporative Cooler</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>111</td>\n",
       "      <td>geothermal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>112</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>113</td>\n",
       "      <td>Air Con</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Cust_Number      Cooling_System\n",
       "0           101                  AC\n",
       "1           102           Heat Pump\n",
       "2           103             Air Con\n",
       "3           104    Air Conditioning\n",
       "4           105                 Fan\n",
       "5           106                None\n",
       "6           107          EvapCooler\n",
       "7           108                None\n",
       "8           109                  EC\n",
       "9           110  Evaporative Cooler\n",
       "10          111          geothermal\n",
       "11          112                   8\n",
       "12          113             Air Con"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas\n",
    "\n",
    "# creating a data frame from scratch - list of lists\n",
    "\n",
    "data = [ \n",
    "         [101, 'AC'], \n",
    "         [102, 'Heat Pump'], \n",
    "         [103, 'Air Con'], \n",
    "         [104, 'Air Conditioning'],\n",
    "         [105, 'Fan'],\n",
    "         [106, 'None'],\n",
    "         [107, 'EvapCooler'],\n",
    "         [108, None],\n",
    "         [109, 'EC'],\n",
    "         [110, 'Evaporative Cooler'],\n",
    "         [111, 'geothermal'],\n",
    "         [112, 8],\n",
    "         [113, 'Air Con']\n",
    "       ]\n",
    "\n",
    "# create a data frame with column names - list of lists\n",
    "\n",
    "col_names = ['Cust_Number', 'Cooling_System']\n",
    "df = pandas.DataFrame(data, columns=col_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cust_Number</th>\n",
       "      <th>Cooling_System</th>\n",
       "      <th>cooling_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>101</td>\n",
       "      <td>AC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>102</td>\n",
       "      <td>Heat Pump</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>103</td>\n",
       "      <td>Air Con</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>104</td>\n",
       "      <td>Air Conditioning</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>105</td>\n",
       "      <td>Fan</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>107</td>\n",
       "      <td>EvapCooler</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>109</td>\n",
       "      <td>EC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>110</td>\n",
       "      <td>Evaporative Cooler</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>113</td>\n",
       "      <td>Air Con</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Cust_Number      Cooling_System  cooling_type\n",
       "0           101                  AC             1\n",
       "1           102           Heat Pump             2\n",
       "2           103             Air Con             1\n",
       "3           104    Air Conditioning             1\n",
       "4           105                 Fan             4\n",
       "6           107          EvapCooler             3\n",
       "8           109                  EC             3\n",
       "9           110  Evaporative Cooler             3\n",
       "12          113             Air Con             1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2b) Your code here\n",
    "import numpy as np\n",
    "CoolSys = { 'Air Conditioning': 1,'Air Con': 1,'AC' : 1,\n",
    "           'Heat Pump': 2, 'HP': 2,\n",
    "           'Evaporative Cooler': 3, 'EvapCooler': 3, 'EC': 3,\n",
    "           'Fan': 4 \n",
    "}\n",
    "\n",
    "df['cooling_type']= np.int8(df['Cooling_System'].map(CoolSys))\n",
    "df[df['cooling_type'] != 0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "c) From the dataframe below, use groupby in Pandas to show how many total delegates were obtained grouped by state. Print this out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>delegates</th>\n",
       "      <th>color</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>marco</td>\n",
       "      <td>165</td>\n",
       "      <td>blue</td>\n",
       "      <td>FL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>jeb</td>\n",
       "      <td>35</td>\n",
       "      <td>red</td>\n",
       "      <td>FL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chris</td>\n",
       "      <td>0</td>\n",
       "      <td>white</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>donald</td>\n",
       "      <td>1543</td>\n",
       "      <td>white</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ted</td>\n",
       "      <td>559</td>\n",
       "      <td>blue</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>john</td>\n",
       "      <td>161</td>\n",
       "      <td>red</td>\n",
       "      <td>OH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     name  delegates  color state\n",
       "0   marco        165   blue    FL\n",
       "1     jeb         35    red    FL\n",
       "2   chris          0  white    NY\n",
       "3  donald       1543  white    NY\n",
       "4     ted        559   blue    NY\n",
       "5    john        161    red    OH"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas\n",
    "\n",
    "# creating a data frame from scratch - list of lists\n",
    "\n",
    "data = [ ['marco', 165, 'blue', 'FL'], \n",
    "         ['jeb', 35, 'red', 'FL'], \n",
    "         ['chris', 0, 'white', 'NY'], \n",
    "         ['donald', 1543, 'white', 'NY'],\n",
    "         ['ted', 559, 'blue', 'NY'],\n",
    "         ['john', 161, 'red', 'OH']\n",
    "       ]\n",
    "\n",
    "# create a data frame with column names - list of lists\n",
    "\n",
    "col_names = ['name', 'delegates', 'color', 'state']\n",
    "df = pandas.DataFrame(data, columns=col_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "state\n",
       "FL     200\n",
       "NY    2102\n",
       "OH     161\n",
       "Name: delegates, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2c) Your code here\n",
    "df.groupby('state')['delegates'].sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## 3: Bitcoin coding problem (20 points):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Bitcoin\n",
    "\n",
    "Consider a record of a one-time investment in bitcoin with value of that investment tracked monthly, provided as an (ordered) tuple of dictionaries, where each dictionary comprises one key for the month and corresponding one value for the value of the investment as of that date. (For example 31 Jan 2018 was $1200 as shown in `data` below).\n",
    "\n",
    "The initial investment is made on 1 Jan 2018 and was for $1000 (labelled 'Initial' in the `data` below).\n",
    "\n",
    "Fill in the Python function (named `bitcoin_invest` below) to take such a record of any length (the below `data` is only a sample but the full data will be in the same format), and return a dataframe comprising a row for each month with columns for date, balance, and return. **Print out/display this dataframe.**\n",
    "\n",
    "Also, visualize the record as two vertically arranged plots (there are a few ways to do this, keep it consistent over both plots). \n",
    "- The top plot should show a line plot of balance vs. month \n",
    "- The bottom plot should show a bar plot of return vs. month, with a black horizontal line at return=0, and bars color-coded such that positive returns are green and negative returns are red.  \n",
    "- The two plots' horizontal axes should align. Demonstrate that your code works by applying it to `data`.\n",
    "\n",
    "Some definitions: <br>\n",
    "- The gain for each period is the end balance minus the start balance. \n",
    "- The growth factor for each period is the end balance divided by the start balance.  \n",
    "- The return for each period is the growth factor minus 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'Initial': 1000},\n",
       " {'Jan 2018': 1200},\n",
       " {'Feb 2018': 1400},\n",
       " {'Mar 2018': 700},\n",
       " {'Apr 2018': 800},\n",
       " {'May 2018': 500})"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = ({\"Initial\": 1000}, {\"Jan 2018\":1200},{\"Feb 2018\":1400},{\"Mar 2018\":700},{\"Apr 2018\":800},{\"May 2018\":500})\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# 3) Your code here\n",
    "import pandas as pd\n",
    "\n",
    "def bitcoin_invest(monthly_data):\n",
    "    df = pd.DataFrame(monthly_data).max().reset_index()\n",
    "    df.columns = ['date', 'balance']\n",
    "    df['lag'] = df.balance.shift(1)\n",
    "    df['gain'] = df['balance'] - df['lag']\n",
    "    df['return'] = df['gain'] / df['lag']\n",
    "    df.drop(['gain', 'lag'], inplace=True, axis=1)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>balance</th>\n",
       "      <th>return</th>\n",
       "      <th>positive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Initial</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Jan 2018</td>\n",
       "      <td>1200.0</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Feb 2018</td>\n",
       "      <td>1400.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mar 2018</td>\n",
       "      <td>700.0</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Apr 2018</td>\n",
       "      <td>800.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>May 2018</td>\n",
       "      <td>500.0</td>\n",
       "      <td>-0.375000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       date  balance    return  positive\n",
       "0   Initial   1000.0       NaN     False\n",
       "1  Jan 2018   1200.0  0.200000      True\n",
       "2  Feb 2018   1400.0  0.166667      True\n",
       "3  Mar 2018    700.0 -0.500000     False\n",
       "4  Apr 2018    800.0  0.142857      True\n",
       "5  May 2018    500.0 -0.375000     False"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "monthly = bitcoin_invest(data)\n",
    "monthly['positive'] = monthly['return'] > 0\n",
    "monthly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAk4UlEQVR4nO3deZwUxf3/8deHUzxQBCTI4aJBIxACsuKteBP0G8QrgEYElXjEiL8YRTFeicZoQhKjxqA50IBXFCUGg2gkmgjqEjlFBAUVRQ6PeKPA5/dH1WbbYY/Z3VlmZvv9fDzmMT3VPd1V27OfqamurjJ3R0RE0qFJvjMgIiJbjoK+iEiKKOiLiKSIgr6ISIoo6IuIpEizfGegJu3atfOSkpJ8Z0NEpKjMmTNnnbu3z0wv+KBfUlJCWVlZvrMhIlJUzOy1ytLVvCMikiIK+iIiKaKgL4XFHebOhR/+EHr1gilT8p0jkUal4Nv0JSWWL4fJk2HSJFi8GJo1g/bt4dRTYdYs6N073zkUaRRU05f8WbsWbrkFDjgAdt0VLr8c2raF3/4WVq2COXNg++3huOPgnXfynVuRRkE1fdmyPv4YHnoo1OqnT4eNG6FnT7juOhg2DDK7506ZAgcfDEOHwqOPhl8AIlJn+g+ShvfFFzBjRmi6eegh+OQT6NIFLroIhg+vvulmn33gtttg1Ci45BL4xS+2WLZFGiMFfWkY7vDMM6FGf999sG4dtGkT2uhPOQUOPBCaZNm6OHIk/Oc/MH489O0b9iEidaKgL7n14ouhRj95MqxYAVttBd/6Vgj0AwdCixZ12+/48bBgAZx1Fuy5J/Trl9Nsi6SFgr7U38qVcPfdIdjPmxdq8EccAVdfDUOGwHbb1f8YzZvD/fdDaWm4sFtWBh061H+/IimjoC9189578Je/hED/1FOhOad/f/j1r+Hb326YgNy+fbiwe+CBcNJJ8Pjjdf/lIJJS6rIp2fv001DbHjIEvvIVGD06dK286ip4+WV49ln4/vcbtga+115wxx3w9NNw4YUNdxyRRko1fanexo3wj3+ENvoHHoAPP4SOHeG880LPm379wGzL5mn4cHjhBfj5z8OF3TPP3LLHFyliCvqyOffQZj55MtxzD7z9NrRuDSeeGALuoYdC06b5zeP118P8+XDuudCjB+y/f37zI1IkFPSlwtKlIdBPnhyaa1q0gGOOCYH+mGOgVat857BC06bh4vHee8MJJ4QvqU6d8p0rkYJXY5u+mf3BzNaY2cJE2o1m9pKZzTezKWa2Q0wvMbNPzWxufNyWeE8/M1tgZsvM7CazLd0mIJV6++1w8bV/f9h999DjplMnuP32sO7BB0MNv5ACfrkdd4SHHw5NTscfD599lu8ciRS8bC7k/gkYmJE2A+jl7r2Bl4FLE+tecfc+8XF2Iv23wGige3xk7lO2lA8+gIkT4aijQoAfMybcNXvjjfD666EN/8wzw81Uha5XL7jzTnjuudDU457vHIkUtBqDvrs/BbybkfaYu2+IL2cDnavbh5l1BFq7+yx3d+BO4Lg65Vjq5vPPQ6345JND75rTT4dly+DSS2HRonBh9KKLoHO1p7IwHX88/OhH8Mc/hgHcRKRKuWjTHwXcm3jdzcxeAD4ALnf3p4FOwMrENitjWqXMbDThVwFdu3bNQRZTatOm0LVx8uTQ1fK996BdOzjjjHCH7L77bvmeNw3lqqvCOPxjxoTa/4AB+c2PSIGqV9A3s3HABmBSTFoFdHX3d8ysH/CQmfUEKossVf4Od/cJwASA0tJS/V6vDffQq2Xy5HCh8403YJttwl2sp5wS7pRt3jzfucy9Jk3gz38OA7SddFK4sLvLLvnOlUjBqXPQN7MRwLHA4bHJBndfD6yPy3PM7BVgd0LNPtlu0Bl4q67HlkqsWFExFMKiRWEI4qOPhp/9LIx9s802+c5hw2vdOozi2b9/uIHsX/+CrbfOd65ECkqd7sg1s4HAJcC33P2TRHp7M2sal3clXLB91d1XAR+a2b6x185pwMP1zn3arVsXJhw58EDo1g0uuwx22AFuvTXcKfvII2GM+jQE/HJ77BF+5cydGy5G68KuyJfUWNM3s7uBAUA7M1sJXEnordMSmBF7Xs6OPXUOBq4xsw3ARuBsdy+/CHwOoSdQK+DR+JDa+vhjmDo11OinT4cNG8LNSddeG/rTZ05CkkbHHAM/+QmMGxeGbbjoonznSKRgmBd4Tai0tNTLysrynY38+uKLMLhY+SQkH38cetkMGxba6Xv3bjwXZHPFPfRUevBBmDYtNHWJpIiZzXH30sx03ZFbqNxh9uwQ6O+7L8wn26ZNCPLDh8NBB2U/CUkamYUunEuWhKkWn38evvrVfOdKJO8U9AvN4sUVk5AsX14xCcnw4WESkpYt853D4rHttuGX0d57h95Ls2blZmx/kSKmoF8I3nyzoufN3LmhBn/44XDllaEXSuvW+c5h8dp1V7j33tC8M2JEmANAv5AkxfTpz5f33w/jwh92WJgk/Ic/DP3nf/Wr8CXw2GMhSCng198RR4RhmKdMCRe8RVJMNf0t6bPPQjfKyZPhb38LQyN07x5q9MOHh2VpGGPGhMnVr7gCvvGN0GQmkkIK+g1t40Z48smKSUg++CDMOnXuueGibD4mIUkjM5gwIVwzOfXUMMvXnnvmO1ciW5yCfkNwhzlzKiYhWbUqXEA84YQQ6AthEpI0atUqNPGUlsLgwWFkzh12yHeuRLYoBf1cWrasYhKSJUvCJCSDBoVAX2iTkKRVly7hYu5hh4XzMnWqvoAlVXQht75Wr4abbgoDfXXvHkZ77NixYhKSKVMKdxKStDrooHDOpk0LbfwiKaKafl18+GEI5pMmhTtlN22CPn3ghhvCXbLFOCZ92px9driwe9114dyddFK+cySyRSjoZ+vzz+Hvfw+BfurU0BOnpATGjg3NBD165DuHUhtmcPPNYUTS008PA7X17p3vXIk0OAX96mzaFIbnnTTpy5OQjBoVAv1++6nnTTFr2TL0qCotDXfsPv88tG2b71yJNCgF/crMnx8CffkkJFtvHe6MHT4cjjyycU5CklYdO4ZB2Q4+GL797fBrrpn+LaTx0qe73GuvVfS8Wbgw9Og4+mi4/vrQvS9NY9KnzT77wG23hV9wF18M48fnO0ciDSbdQX/dutBsM3lyaMYB2H//MLn2SSdB+/b5zZ9sOSNHhsnhf/lL6NsXvvOdfOdIpEGkL+iXT0IyeXL4KZ+chGTYsDADlaTTL34BCxbAWWeFu3VLNxuKXKTo1dhP38z+YGZrzGxhIm1HM5thZkvjc5vEukvNbJmZLTGzoxPp/cxsQVx3U5w2ccvYsAEefTTU3jp0CG3zc+fChReG54ULw1SDCvjp1rx5mLugQ4dwDWf16nznSCTnsrk560/AwIy0scAT7t4deCK+xsx6AEOBnvE9t5bPmQv8FhhNmDe3eyX7zC33MH76+efDzjuHO2MfeSQE/JkzQxv+DTeEwbfUA0fKtW8fxuB/551wU93nn+c7RyI5VWPQd/engHczkgcDE+PyROC4RPo97r7e3ZcDy4D+ZtYRaO3uszzMz3hn4j255x7mRt1//zB88YAB4R/57bfDoFuHHKIx1aVqffvC738frvOMGZPv3IjkVF3b9Du4+yoAd19lZjvF9E7A7MR2K2PaF3E5M71SZjaa8KuArl271j53ZmEkxQsugOOP15j0UnvDhoULuzfeGL4Ezjor3zkSyYlcX8itrJ3Eq0mvlLtPACZAmBi9Tjn5wQ/q9DaR//npT2HePDjvPOjZM/xyFClydW3jWB2bbIjPa2L6SqBLYrvOwFsxvXMl6SKFq2nTMDR2165hWOw338x3jkTqra5BfyowIi6PAB5OpA81s5Zm1o1wwfa52BT0oZntG3vtnJZ4j0jhatMGHn44DLJ3/PFhzCWRIpZNl827gVnAHma20szOAK4HjjSzpcCR8TXuvgi4D3gR+DtwnrtvjLs6B7iDcHH3FeDRHJdFpGH07Al33RUmXTnnnNBRQKRImRf4B7i0tNTLysrynQ2RMJfxNdeEsfjPPz/fuRGplpnNcffN7jBUv0WRbF15ZZhQ/cILw70eIkVIQV8kW02ahGae7t3D2EyvvZbvHInUmoK+SG20bh0u7H7xRRiD/5NP8p0jkVpR0Beprd13DwP2zZsHZ5yhC7tSVBT0Repi0KAwMus998DPf57v3IhkTUFfpK7Gjg1t+2PHwvTp+c6NSFYU9EXqygz++Efo1QuGDoVly/KdI5EaKeiL1Mc224QRXJs0CdNqfvhhvnMkUi0FfZH66tYtTL6yZAmcdhps2pTvHIlUSUFfJBcOPzxc0H3ooXCBV6RAKeiL5MoFF4QpOa+4IszDLFKAFPRFcsUMfve7MKH6qafC4sX5zpHIZhT0RXKpVSt48MHwPHgwvP9+vnMk8iUK+iK51qUL/OUvsHw5DB8OGzfW/J5it3EjvPwyPPBAGJju+OPDGEXt2sFVV8EHH+Q7hxLlerpEEQE46CD4zW/C+Ps/+hFcd12+c5Q7a9bAggXhMX9+eF60CD79NKxv0iQE/D594PPP4eqr4ZZbYNw4OPts2GqrvGY/7TSevkhDcYfvfhduvx3uvRdOPjnfOaqdTz+FF1/8cnBfsABWr67YpkMH+PrXw6N37/Dco0do3ir3/PNw2WXw+ONh6smrrw4XvJs23fJlSpGqxtOvc9A3sz2AexNJuwJXADsAZwFrY/pl7j4tvudS4AxgI/B9d6/x3nUFfSlq69fDYYfB3LnwzDPwjW/kO0eb27QpNEVlBvelSyvuOWjVKswglgzuX/867LRT9sd5/HG49FIoKwtfDNdeG657mDVMuVIu50E/Y+dNgTeBfYCRwEfu/vOMbXoAdwP9gZ2Bx4HdE9MpVkpBX4reqlWhR0+LFiHgtW2bv7y8887mwX3hQvj447DeDHbbbfPgvttuuamZu4cL3ePGhZvZ9tkHrr8eBgyo/77lS6oK+rlq0z8ceMXdX7Oqv7UHA/e4+3pguZktI3wBzMpRHkQKU8eOMGVKaOc/+eQwOFuzBr6ctn596DKaDPDz54cvoHJt24bAfsYZFQG+Z88wtERDMYMTTgg1/IkTw0XfQw+Fo4+Gn/4U+vZtuGMLkLugP5RQiy/3PTM7DSgDfuDu7wGdgNmJbVbGtM2Y2WhgNEDXrl1zlEWRPOrfP/ThHzkSLr4Yxo/PzX7dwwxembX3JUsqeg21bBmaU4488su19698JX9NK82ahS+b4cPh1lvDhe699goD1/34x/DVr+YnXylQ7+YdM2sBvAX0dPfVZtYBWAc48GOgo7uPMrNbgFnu/uf4vt8D09z9ger2r+YdaVQuuCBMrD5xYhinpzbef3/z4L5gwZcHeevWbfMLq927N/wvi/r673/DMBbjx4ceP2eeGXo97bxzvnNWtBqsTd/MBgPnuftRlawrAR5x917xIi7u/tO4bjpwlbtX27yjoC+NyhdfhKaMZ56Bf/0rtPVn+vzzUFNPBvf582Hlyopt2rTZPLj36gXbbbflytIQ3n4bfvKT8KuoefPwJXnxxaG8UisNGfTvAaa7+x/j647uviouXwjs4+5DzawnMJmKC7lPAN11IVdSZ+1a2Hvv0PzyyCMhmCeD+0svwYYNYdvmzWHPPTcP8J06Ne5eL6++GsYwmjwZtt8+TFRz/vmw9db5zlnRaJCgb2ZbA28Au7r7f2PaXUAfQvPOCuC7iS+BccAoYAMwxt0frekYCvrSKM2dC/vvX3FDE4Q+7JnBfY89QuBPq3nzQk+fv/0tXBC/8koYNSrdf5MsNWiXzYakoC+N1pw54calXr3CY4cd8p2jwvX006GP/7//HS7y/uQnYarKJhpJpipVBX39xUTypV+/MCzBgQcq4NfkoINC4P/rX8MwDkOHhush06eHHkySNQV9ESkOZnDssaFp7K674L33YODAcMfz7Nk1vl0CBX0RKS5Nm4b5CpYsCYPavfgi7LcfDBkSlqVaCvoiUpxatIDvfQ9eeSXc0PXEE+Hi98iR4YY1qZSCvogUt223hcsvD908L7wQ7r4bdt89LK9dW/P7U0ZBX0Qah3btwl29S5eGoZtvuikMFHfNNV++aznlFPRFpHHp0gXuuCOMHnrUUaFv/267hS+B9evznbu8U9AXkcZpzz3DtJXPPhva+i+4INzsdued6ZjCsgoK+iLSuPXvHyZweeyx0AQ0YkSYzGbq1FT28VfQF5HGzywMLf3883D//WHgu8GD4YAD4Kmn8p27LUpBX0TSwwxOPDFM5H777fD663DIITBoULjpKwUU9EUkfZo1C2P2L10KN9wQ7ujt2zdM6rJsWb5z16AU9EUkvVq1gh/+MPTxv+wyePjhcAH43HO/PLVkI6KgLyKyww5w7bWhlj96dGj62W238EXw/vv5zl1OKeiLiJTr2BFuuSVMZDNkSJisfdddQxNQcu6DIqagLyKSabfdYNIkeOGFMJjbJZeEcfxvv71iVrMiVa+gb2YrzGyBmc01s7KYtqOZzTCzpfG5TWL7S81smZktMbOj65t5EZEG1adPmLXrn/+EkpLQ9NOzZ+j2uWlTvnNXJ7mo6R/q7n0SM7SMBZ5w9+6EeXDHAphZD2Ao0BMYCNxqZk1zcHwRkYZ18MFhIvupU8PoniefHG76mjGj6G7waojmncHAxLg8ETgukX6Pu6939+XAMsIk6SIihc8M/u//Qn/+O++EdesqxvYpIvUN+g48ZmZzzGx0TOtQPhF6fN4ppnciTKJebmVM24yZjTazMjMrW6uhUUWkkDRtGkbxXLIkDOJ28sn5zlGtNKvn+w9w97fMbCdghpm9VM22Vklapb+L3H0CMAHCxOj1zKOISO61bAnnn5/vXNRavWr67v5WfF4DTCE016w2s44A8XlN3Hwl0CXx9s7AW/U5voiI1E6dg76ZbWNm25UvA0cBC4GpwIi42Qjg4bg8FRhqZi3NrBvQHXiurscXEZHaM6/jlWcz25VQu4fQTDTZ3a81s7bAfUBX4HXgJHd/N75nHDAK2ACMcfdHszjOWqCuE162A9bV8b3FoLGXDxp/GVW+4leoZdzF3dtnJtY56BcDMytLdCVtdBp7+aDxl1HlK37FVkbdkSsikiIK+iIiKdLYg/6EfGeggTX28kHjL6PKV/yKqoyNuk1fRES+rLHX9EVEJEFBX0QkRRT0RURSREFfRCRFFPRFRFJEQV9EJEUU9EVEUkRBX0QkRRT0RURSREFfRCRFFPRFRFJEQV9EJEUU9EVEUkRBX0QkRZrlOwM1adeunZeUlOQ7GyIiRWXOnDnrKpsjN2dB38wGAr8GmgJ3uPv1GetPAS6JLz8CznH3eTXtt6SkhLKyslxlU0QkFczstcrSc9K8Y2ZNgVuAbwI9gGFm1iNjs+XAIe7eG/gxRTbbjIhIY5CrNv3+wDJ3f9XdPwfuAQYnN3D3Z9z9vfhyNtA5R8cWEZEs5ap5pxPwRuL1SmCfarY/A3i0qpVmNhoYDdC1a9dc5C+v7GrLdxZqza/UNJoijVGuavqVRbVKo4aZHUoI+pdUth7A3Se4e6m7l7Zvv9l1CBERqaNc1fRXAl0SrzsDb2VuZGa9gTuAb7r7Ozk6toiIZClXNf3nge5m1s3MWgBDganJDcysK/Ag8B13fzlHxxURkVrISU3f3TeY2feA6YQum39w90VmdnZcfxtwBdAWuNXMADa4e2kuji8iItnJWT99d58GTMtIuy2xfCZwZq6OJ4VDF6pFioeGYRARSREFfRGRFCn4sXdEpOGpiS49VNMXEUkRBX0RkRRR0BcRSREFfRGRFFHQFxFJEQV9EZEUUdAXEUkRBX0RkRRR0BcRSREFfRGRFFHQFxFJEQV9EZEUUdAXEUkRBX0RkRTJWdA3s4FmtsTMlpnZ2ErWm5ndFNfPN7O9cnVsERHJTk6Cvpk1BW4Bvgn0AIaZWY+Mzb4JdI+P0cBvc3FsERHJXq4mUekPLHP3VwHM7B5gMPBiYpvBwJ3u7sBsM9vBzDq6+6rqdrxkyRIGDBiQo2zmyYp8Z6D2Bjw5IPuNVzRULhpOrcqXBivynYHa0zmsm1wF/U7AG4nXK4F9stimE7BZ0Dez0YRfA7Rs2TJHWcyfQ0oOyXcWGlRjLx///Ge+c1A3h2R/XnQOC1Atzl9t5CroVzbXWuZcZtlsExLdJwATAEpLS33mzJn1ypxIvVjxTSUIgP5vKhTjOazn+bMqypyrC7krgS6J152Bt+qwjYiINKBcBf3nge5m1s3MWgBDgakZ20wFTou9ePYF/ltTe76IiORWTpp33H2DmX0PmA40Bf7g7ovM7Oy4/jZgGjAIWAZ8AozMxbFFRCR7uWrTx92nEQJ7Mu22xLID5+XqeCIiUnu6I1dEJEUU9EVEUkRBX0QkRRT0RURSREFfRCRFFPRFRFJEQV9EJEUU9EVEUkRBX0QkRRT0RURSREFfRCRFFPRFRFJEQV9EJEUU9EVEUkRBX0QkRRT0RURSREFfRCRF6h30zWxHM5thZkvjc5tKtuliZk+a2WIzW2RmF9T3uCIiUnu5qOmPBZ5w9+7AE/F1pg3AD9x9T2Bf4Dwz65GDY4uISC3kIugPBibG5YnAcZkbuPsqd/9PXP4QWAx0ysGxRUSkFnIR9Du4+yoIwR3YqbqNzawE6As8W802o82szMzK1q5dm4MsiogIQLNsNjKzx4GvVLJqXG0OZmbbAg8AY9z9g6q2c/cJwASA0tJSr80xRESkalkFfXc/oqp1ZrbazDq6+yoz6wisqWK75oSAP8ndH6xTbkVEpF5y0bwzFRgRl0cAD2duYGYG/B5Y7O7jc3BMERGpg1wE/euBI81sKXBkfI2Z7Wxm0+I2BwDfAQ4zs7nxMSgHxxYRkVrIqnmnOu7+DnB4JelvAYPi8r8Aq++xRESkfnRHrohIiijoi4ikiIK+iEiKKOiLiKSIgr6ISIoo6IuIpIiCvohIiijoi4ikiIK+iEiKKOiLiKSIgr6ISIoo6IuIpIiCvohIiijoi4ikiIK+iEiKKOiLiKRIvSdRMbMdgXuBEmAFcLK7v1fFtk2BMuBNdz+2vscWEcmKe75zUDByUdMfCzzh7t2BJ+LrqlwALM7BMUVEpA5yEfQHAxPj8kTguMo2MrPOwDHAHTk4poiI1EEugn4Hd18FEJ93qmK7XwEXA5tq2qGZjTazMjMrW7t2bQ6yKCIikGWbvpk9DnylklXjsnz/scAad59jZgNq2t7dJwATAEpLS9UYJyKSI1kFfXc/oqp1ZrbazDq6+yoz6wisqWSzA4BvmdkgYCugtZn92d1PrVOuRUSkTnLRvDMVGBGXRwAPZ27g7pe6e2d3LwGGAv9QwBcR2fJyEfSvB440s6XAkfE1ZrazmU3Lwf5FRCRH6t1P393fAQ6vJP0tYFAl6TOBmfU9roiI1J7uyBURSREFfRGRFFHQFxFJEQV9EZEUUdAXEUkRBX0RkRRR0BcRSZF699MXafQ0Frs0Iqrpi4ikiIK+iEiKKOiLiKSIgr6ISIqYF/hFKjNbC7xWx7e3A9blMDuFprGXDxp/GVW+4leoZdzF3dtnJhZ80K8PMytz99J856OhNPbyQeMvo8pX/IqtjGreERFJEQV9EZEUaexBf0K+M9DAGnv5oPGXUeUrfkVVxkbdpi8iIl/W2Gv6IiKSoKAvIpIiRRP0zeyjLLa5w8x6xOXLMtY9k4tj1Ecu9m9m/8/MXjSz+Wb2hJntklg3wsyWxseIRPr3zGyZmbmZtUukb29mfzWzeWa2yMxG1iNfG81sbuJRUs22M82s2i5uZnakmc0xswXx+bDEun4xfZmZ3WRmFtMPNrP/mNkGMzsxY383xDIuTr4ny7K5md2VeN3MzNaa2SPZ7qOSfXYxsydjfhaZ2QWJdTua2Yx4HmeYWZuY3ja+5yMzuzljf8Pi32S+mf09eZ5rkachsaxfq2u5Evu60cxeivmZYmY7JNZdGs/dEjM7OpF+rZm9kfl/YmZdY7lfiPsbVMu85Pz8JfZVMJ/TrLl7UTyAjxpy+7q+Z0vvHzgU2DounwPcG5d3BF6Nz23icpu4ri9QAqwA2iX2dRnws7jcHngXaNHQZQNmAqU1bNMX2Dku9wLeTKx7DtgPMOBR4JsxvQToDdwJnJjYfn/g30DT+JgFDKhN2YAXgFbx9TeBucAjtdhHs4zXHYG94vJ2wMtAj/j6BmBsXB6bOEfbAAcCZwM3J/cNrCk/t/H9V9XhHN4HPF3H9zbNeH1UeZmBnyXK0AOYB7QEugGvlL8X2Df+XT7K2NcE4JzE+1fU9rNZ3/NXDJ/TbB9FU9MvZ2YDYk3xL7EmMSnxDTrTzErN7HqgVaxxTorrPorP21qoIf8nfgsP3sL5r/T4ZlYSv91vj9/0j5lZq8z3u/uT7v5JfDkb6ByXjwZmuPu77v4eMAMYGN/zgruvqCQ7DmwX/37bEoL+hhyWtZ+Z/TPWgKabWcfE6lPN7BkzW2hm/TfLWMjzW/HlImArM2sZ99Ha3Wd5+E+5EzguvmeFu88HNlVSzq2AFoRg0xxYXcviPAocE5eHAXcnytk/luWF+LxHTD/dzO43s78Cj2WUb5W7/ycufwgsBjrF1YOBiXF5YqJ8H7v7v4DPMvJm8bFNPJetgbeoBTPbFjgAOAMYmkgfYGZPxdr6i2Z2m5k1ies+MrNrzOxZQnBLlu8xdy//LCU/p4OBe9x9vbsvB5YB/eN7Zrv7qkqy57FMANvXtmxRXc7f02bWJ7Hdv82sd0Y5C+1zWqOiC/pRX2AM4Vt/V8KH9X/cfSzwqbv3cfdTMt77GTDE3fci1Jp/0SA/oapW3fG7A7e4e0/gfeCEGvZ1BuHDDCFgvJFYt5KKIFKVm4E9Cf9EC4AL3D3zg5it8i/ZuTFANAd+Q6jJ9AP+AFyb2H4bd98fODeuq84JwAvuvp5QppWJdTWW091nAU8Cq+JjursvrkXZAO4BhprZVoRa2rOJdS8BB7t7X+AK4LrEuv2AEe5+GFWw0BTWN7HPDuXBLz7vVF3G3P0Lwq++BYRz2QP4fdYlC44D/u7uLwPvmtleiXX9gR8AXwd2A46P6dsAC919n/hlVJVR1O9zehWhkrASmAacX2NpNleX83cHcDqAme0OtIzBuiqF8DmtUbEG/efcfWUMUHMJP5eyZcB1ZjYfeJxwIjrkPId1O/5yd58bl+dQTbnM7FSgFLgxsd9MNfXHPZrw99sZ6APcbGatq3tDNcq/ZPu4+xBgD8LP3RlmNhe4nIraHsSalrs/BbS2RJtvkpn1JDQPfLc8qZLNqi2nmX2V8OXWmfD3PszMDs6yXMR8ziecj2GEwJO0PXC/mS0Efgn0TKyb4e7vVpO3bYEHgDHu/kFt8pTYR3NC0O9LOJfzgUtruZthhMBIfB6WWPecu7/q7hsJ5+3AmL4x5r26vI0j/HqcVJ5UyWY1fU6HAX9y987AIOCu8l8b2arj+bsfODb+fUcBf6pq/4XyOc1Gsc6ctT6xvJHaleMUQvt1P3f/wsxWEH5SbSnVHT+zXJs17wCY2RHAOOCQWKuAUJMYkNisM6HtvDojgevjz89lZrYc+BqhLbK+DFjk7vtVsT7zH2Czfwgz6wxMAU5z91di8kq+/OXRmZp/7g8BZrt7eRPfo4T246dqeF+mqcDPCX/nton0HwNPuvuQWGufmVj3cVU7i8HkAWCSuz+YWLXazDq6+6rYTLCmhnz1ASj/G5nZfYRrAVkxs7bAYUAvM3NCe7Kb2cVxk6rO1Wfxi6Cq/Y4AjgUOj58xCOevS2KzbM7fGVQ0Vc6KtfV21Px3yVSr8+fun5jZDEKT1MmEStZmCvBzWq1ireln44v4T5Vpe2BNDLiHArtUsk1Dqtfxzawv8DvgW+6e/NBPB44yszYWenscFdOq8zpweNxvB0Lt/NXa5KcaS4D2ZrZf3H/zWBsq9+2YfiDwX3f/b/LNseb/N+BSd/93eXps7vjQzPaNzWKnAQ/XkJfXgUMs9NpoDhxCaEOvrT8A17j7goz07YE34/Lp2ewo5v33wGJ3H5+xeipQ3vtqBDWX702gh5mVj6h4JLUr34nAne6+i7uXuHsXYDkVNfr+ZtYt1q6/DVTXlAOAmQ0ELiF8Tj9JrJpKaGZpaWbdCE2aNVUykp/TPQmVpLXZF+9/6nL+7gBuAp6v7BdbgX5Oq5fN1d5CeBCv6BO+pR9JpN8MnB6XZxJ7hRB+ai0m1KKS729HuCpeRjihi4GS5DYNlP9mwDtVHT8+Fia2v4hKelEQmoRWE5pl5gJTE+tGES6MLQNGJtK/T6h5bCDUNu6I6TsTLjAuABYCp9b3/GSk9SHUUuYRLnKdlThPPwWeicftX8l7LyfUkucmHjvFdaXxfa/E819+Z/nesZwfx7/1opjelPBFuRh4ERifg7L973NIaLd/mdDz4sfE3iWEAHJzFfs8kFBjnp8o36C4ri3wBLA0Pu+YeN8KwgX3j2JZy3v8nB3LNx/4K9C2FuWbCQzMSPs+8NtYzn8A98a/3W1Ak5r+X+Jn8I1E2W5LrBsXz90SYo+WmH5DLNOm+HxVTO8R/7bz4r6O2hLnL7HtS5l/n0L8nGb70DAMW4iZfQO43d0366kiUqjMbABwkbsfm+es5IWZ7Uz4Uvya172TQ0FpzM07BcPMziZcALs833kRkeyY2WmEXj7jGkvABw24JiKSKqrpi4ikiIK+iEiKKOiLiKSIgr6ISIoo6IuIpMj/B1ZAjCftL0biAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(2, 1, sharex=True)\n",
    "\n",
    "c = 'g' if monthly.balance.iloc[-1] >= monthly.balance.iloc[0] else 'r'\n",
    "ax1.plot(\n",
    "    monthly.date, \n",
    "    monthly.balance,\n",
    "    color=c\n",
    ")\n",
    "ax2.bar(\n",
    "    monthly.date, \n",
    "    monthly['return'],\n",
    "    color=monthly.positive.map({True: 'g', False: 'r'})\n",
    ")\n",
    "ax2.axhline(y=0, color='k')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## 4: Clinical disease data (40 pts)\n",
    "\n",
    "Your boss comes to you Monday morning and says “I figured out our next step; we are going to pivot from an online craft store and become a data center for genetic disease information! I found **ClinVar** which is a repository that contains expert curated data, and it is free for the taking. This is a gold mine! Look at the file and tell me what gene and mutation combinations are classified as dangerous.”\n",
    "\n",
    "Make sure that you only give your boss the dangerous mutations and include:\n",
    "\n",
    "1) Gene name\n",
    "\n",
    "2) Mutation ID number\n",
    "\n",
    "3) Mutation Position (chromosome & position)\n",
    "\n",
    "4) Mutation value (reference & alternate bases)\n",
    "\n",
    "5) Clinical significance (CLNSIG)\n",
    "\n",
    "6) Disease that is implicated\n",
    "\n",
    "**Requirements**\n",
    "\n",
    "1) The deliverables are the final result as a dataframe with a short discussion of any specifics. (that is, what data you would present to your boss with the explanation of your results)\n",
    "\n",
    "2) Limit your output to the first 100 harmful mutations and tell your boss how many total harmful mutations were found in the file\n",
    "\n",
    "3) Use the instructor-modified \"clinvar_final.txt\" at this link: https://drive.google.com/file/d/1Zps0YssoJbZHrn6iLte2RDLlgruhAX1s/view?usp=sharing This file was modified to be not exactly the same as 'standard' .vcf file to test your data parsing skills. **This is a large file so do NOT upload it into your github repo!**\n",
    "\n",
    "4) Replace missing values in the dataframe with: 'Not_Given'. Print or display this (including the Not_Given count) for the column `CLNSIG` by using pandas value_counts() function (https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html).\n",
    "\n",
    "5) State in your answer how you define harmful mutations\n",
    "\n",
    "**6) Do your best on getting to above requirements and submit whatever you do before the deadline. If your work is incomplete be sure to describe the blockers that got in your way and how you might get past them (if given more time).**\n",
    "\n",
    "7) You can use as many code blocks as you need. Please clean-up your code and make it readable for the graders!\n",
    "\n",
    "**Hints** \n",
    "* We do not expect you to have any medical knowledge to solve this problem; look at the data, read the documentation provided, and write down your assumptions!\n",
    "\n",
    "* Correct pseudocode will give you partial credit so start with that. \n",
    "\n",
    "* Map out which fields you want to extract: Are they in the same place every time? What strategy will you use to robustly extract and filter your data of interest? How do you plan to handle missing data?\n",
    "\n",
    "* A good way to start is to print out each line, then practice parsing them to see if you can recover the fields of interest\n",
    "\n",
    "* A starting solution for parsing .vcfs can be found here: https://gist.github.com/dceoy/99d976a2c01e7f0ba1c813778f9db744 This solution does **NOT** work due to the changes we've made but can be modified to work. As with any solution that needs modifications, it may take less time to make your own solution!\n",
    "\n",
    "* Filter out junk and lines with no mutation data. Just focus on the data your need to deliver to your boss. \n",
    "\n",
    "* Pandas and NumPy parsers correctly recognize the end of each line in in the ClinVar file.\n",
    "\n",
    "* The unit of observation of this dataset is one row per mutation.\n",
    "\n",
    "* This is similar to a task that one of us tackled at work. You can answer the question with the information provided below or using the (partial) data dictionary file at this link: https://drive.google.com/file/d/1lx9yHdlcqmU_OlHiTUXKC_LQDqYBypH_/view?usp=sharing. Our goal is to see that you can put together a sensible plan, implement a solid parsing strategy, and document and justify the decisions that you made."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### VCF file description (Summarized from version 4.1)\n",
    "\n",
    "```\n",
    "* The VCF specification:\n",
    "\n",
    "VCF is a text file format which contains meta-information lines, a header line, and then data lines each containing information about a position in the genome. The format also can contain genotype information on samples for each position.\n",
    "\n",
    "* Fixed fields:\n",
    "\n",
    "There are 8 fixed fields per record. All data lines are **tab-delimited**. In all cases, missing values are specified with a dot (‘.’). \n",
    "\n",
    "1. CHROM - chromosome number\n",
    "2. POS - position DNA nuceleotide count (bases) along the chromosome\n",
    "3. ID - The unique identifier for each mutation\n",
    "4. REF - reference base(s)\n",
    "5. ALT - alternate base(s)\n",
    "6. FILTER - filter status\n",
    "7. QUAL - quality\n",
    "8. INFO - a semicolon-separated series of keys with values in the format: <key>=<data>\n",
    "\n",
    "```\n",
    "### Applicable INFO field specifications\n",
    "\n",
    "```\n",
    "GENEINFO = <Gene name>\n",
    "CLNSIG =  <Clinical significance>\n",
    "CLNDN = <Disease name>\n",
    "```\n",
    "\n",
    "### Sample ClinVar data (vcf file format - not exactly the same as the file to download!)\n",
    "\n",
    "```\n",
    "##fileformat=VCFv4.1\n",
    "##fileDate=2019-03-19\n",
    "##source=ClinVar\n",
    "##reference=GRCh38\t\t\t\t\t\t\t\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n",
    "1\t949523\trs786201005\tC\tT\t.\t.\tGENEINFO=ISG15;CLNSIG=5\n",
    "1\t949696\trs672601345\tC\tCG\t.\t.\tGENEINFO=ISG15;CLNSIG=5;CLNDN=Cancer\n",
    "1\t949739\trs672601312\tG\tT\t.\t.\tGENEINFO=ISG15;CLNDBN=Cancer\n",
    "1\t955597\trs115173026\tG\tT\t.\t.\tGENEINFO=AGRN;CLNSIG=2; CLNDN=Cancer\n",
    "1\t955619\trs201073369\tG\tC\t.\t.\tGENEINFO=AGG;CLNDN=Heart_dis \n",
    "1\t957640\trs6657048\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=3;CLNDN=Heart_dis \n",
    "1\t976059\trs544749044\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=0;CLNDN=Heart_dis \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# 4) Your code here - can use as many code blocks as you would like\n",
    "import io\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('#')]\n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'CHROM': str, 'POS': str, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Reading the file into a dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>542074</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>542075</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS      ID REF ALT FILTER QUAL  \\\n",
       "0     1  1014O42  475283   G   A      .    .   \n",
       "1     1  1O14122  542074   C   T      .    .   \n",
       "2     1  1014143  183381   C   T      .    .   \n",
       "3     1  1014179  542075   C   T      .    .   \n",
       "4     1  1014217  475278   C   T      .    .   \n",
       "\n",
       "                                                INFO  \n",
       "0  AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...  \n",
       "1  AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...  \n",
       "2  ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "3  ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "4  AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PATH = 'clinvar_final.txt'\n",
    "l = read_vcf(PATH)\n",
    "l.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM          0\n",
       "POS            0\n",
       "ID             0\n",
       "REF            0\n",
       "ALT           86\n",
       "FILTER    102321\n",
       "QUAL      102321\n",
       "INFO           0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Which rows have the most missing values\n",
    "l[l == '.'].notnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "We first need to begin by parsing the key=data combinations for the `INFO` column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# Create a dictionary list of the column INFO\n",
    "dict_lst = l['INFO'].str.split(';').apply(\n",
    "    lambda r: dict([\n",
    "        item.split('=') for item in r\n",
    "    ])\n",
    ")\n",
    "\n",
    "# Convert the INFO dictionary list to a DataFrame\n",
    "info_cols = pd.DataFrame(\n",
    "    dict_lst.values.tolist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>ALLELEID</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>CLNHGVS</th>\n",
       "      <th>CLNREVSTAT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNVC</th>\n",
       "      <th>...</th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>MC</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>RS</th>\n",
       "      <th>CLNVI</th>\n",
       "      <th>CLNSIGCONF</th>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <th>CLNDNINCL</th>\n",
       "      <th>CLNSIGINCL</th>\n",
       "      <th>DBVARID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.00546</td>\n",
       "      <td>0.00165</td>\n",
       "      <td>0.00619</td>\n",
       "      <td>446939</td>\n",
       "      <td>MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>NC_000001.11:g.1014042G&gt;A</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>143888043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.00015</td>\n",
       "      <td>0.00010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>514926</td>\n",
       "      <td>MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>NC_000001.11:g.1014122C&gt;T</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>150861311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>181485</td>\n",
       "      <td>MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>NC_000001.11:g.1014143C&gt;T</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001587|nonsense</td>\n",
       "      <td>1</td>\n",
       "      <td>786201005</td>\n",
       "      <td>OMIM_Allelic_Variant:147571.0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>514896</td>\n",
       "      <td>MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>NC_000001.11:g.1014179C&gt;T</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>1553169766</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.00515</td>\n",
       "      <td>0.00831</td>\n",
       "      <td>0.00339</td>\n",
       "      <td>446987</td>\n",
       "      <td>MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>NC_000001.11:g.1014217C&gt;T</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001819|synonymous_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>61766284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AF_ESP  AF_EXAC   AF_TGP ALLELEID  \\\n",
       "0  0.00546  0.00165  0.00619   446939   \n",
       "1  0.00015  0.00010      NaN   514926   \n",
       "2      NaN      NaN      NaN   181485   \n",
       "3      NaN      NaN      NaN   514896   \n",
       "4  0.00515  0.00831  0.00339   446987   \n",
       "\n",
       "                                           CLNDISDB  \\\n",
       "0  MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563   \n",
       "1  MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563   \n",
       "2  MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563   \n",
       "3  MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563   \n",
       "4  MedGen:C4015293,OMIM:616126,Orphanet:ORPHA319563   \n",
       "\n",
       "                                               CLNDN  \\\n",
       "0  Immunodeficiency_38_with_basal_ganglia_calcifi...   \n",
       "1  Immunodeficiency_38_with_basal_ganglia_calcifi...   \n",
       "2  Immunodeficiency_38_with_basal_ganglia_calcifi...   \n",
       "3  Immunodeficiency_38_with_basal_ganglia_calcifi...   \n",
       "4  Immunodeficiency_38_with_basal_ganglia_calcifi...   \n",
       "\n",
       "                     CLNHGVS                           CLNREVSTAT  \\\n",
       "0  NC_000001.11:g.1014042G>A  criteria_provided,_single_submitter   \n",
       "1  NC_000001.11:g.1014122C>T  criteria_provided,_single_submitter   \n",
       "2  NC_000001.11:g.1014143C>T       no_assertion_criteria_provided   \n",
       "3  NC_000001.11:g.1014179C>T  criteria_provided,_single_submitter   \n",
       "4  NC_000001.11:g.1014217C>T  criteria_provided,_single_submitter   \n",
       "\n",
       "                   CLNSIG                      CLNVC  ...    GENEINFO  \\\n",
       "0                  Benign  single_nucleotide_variant  ...  ISG15:9636   \n",
       "1  Uncertain_significance  single_nucleotide_variant  ...  ISG15:9636   \n",
       "2              Pathogenic  single_nucleotide_variant  ...  ISG15:9636   \n",
       "3  Uncertain_significance  single_nucleotide_variant  ...  ISG15:9636   \n",
       "4                  Benign  single_nucleotide_variant  ...  ISG15:9636   \n",
       "\n",
       "                              MC ORIGIN          RS  \\\n",
       "0    SO:0001583|missense_variant      1   143888043   \n",
       "1    SO:0001583|missense_variant      1   150861311   \n",
       "2            SO:0001587|nonsense      1   786201005   \n",
       "3    SO:0001583|missense_variant      1  1553169766   \n",
       "4  SO:0001819|synonymous_variant      1    61766284   \n",
       "\n",
       "                              CLNVI CLNSIGCONF CLNDISDBINCL CLNDNINCL  \\\n",
       "0                               NaN        NaN          NaN       NaN   \n",
       "1                               NaN        NaN          NaN       NaN   \n",
       "2  OMIM_Allelic_Variant:147571.0003        NaN          NaN       NaN   \n",
       "3                               NaN        NaN          NaN       NaN   \n",
       "4                               NaN        NaN          NaN       NaN   \n",
       "\n",
       "  CLNSIGINCL DBVARID  \n",
       "0        NaN     NaN  \n",
       "1        NaN     NaN  \n",
       "2        NaN     NaN  \n",
       "3        NaN     NaN  \n",
       "4        NaN     NaN  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_cols.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CHROM', 'POS', 'ID', 'REF', 'ALT', 'FILTER', 'QUAL', 'INFO'], dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Applicable INFO field specifications\n",
    "\n",
    "```\n",
    "GENEINFO = <Gene name>\n",
    "CLNSIG =  <Clinical significance>\n",
    "CLNDN = <Disease name>\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Replace missing values in the dataframe with: 'Not_Given'. Print or display this (including the Not_Given count) for the column `CLNSIG` by using pandas value_counts() function "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>...</th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>MC</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>RS</th>\n",
       "      <th>CLNVI</th>\n",
       "      <th>CLNSIGCONF</th>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <th>CLNDNINCL</th>\n",
       "      <th>CLNSIGINCL</th>\n",
       "      <th>DBVARID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>0.00546</td>\n",
       "      <td>0.00165</td>\n",
       "      <td>0.00619</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>143888043</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>542074</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>0.00015</td>\n",
       "      <td>0.00010</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>150861311</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001587|nonsense</td>\n",
       "      <td>1</td>\n",
       "      <td>786201005</td>\n",
       "      <td>OMIM_Allelic_Variant:147571.0003</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>542075</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>1553169766</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>0.00515</td>\n",
       "      <td>0.00831</td>\n",
       "      <td>0.00339</td>\n",
       "      <td>...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>SO:0001819|synonymous_variant</td>\n",
       "      <td>1</td>\n",
       "      <td>61766284</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS      ID REF ALT     FILTER       QUAL     AF_ESP    AF_EXAC  \\\n",
       "0     1  1014O42  475283   G   A  Not_Given  Not_Given    0.00546    0.00165   \n",
       "1     1  1O14122  542074   C   T  Not_Given  Not_Given    0.00015    0.00010   \n",
       "2     1  1014143  183381   C   T  Not_Given  Not_Given  Not_Given  Not_Given   \n",
       "3     1  1014179  542075   C   T  Not_Given  Not_Given  Not_Given  Not_Given   \n",
       "4     1  1014217  475278   C   T  Not_Given  Not_Given    0.00515    0.00831   \n",
       "\n",
       "      AF_TGP  ...    GENEINFO                             MC ORIGIN  \\\n",
       "0    0.00619  ...  ISG15:9636    SO:0001583|missense_variant      1   \n",
       "1  Not_Given  ...  ISG15:9636    SO:0001583|missense_variant      1   \n",
       "2  Not_Given  ...  ISG15:9636            SO:0001587|nonsense      1   \n",
       "3  Not_Given  ...  ISG15:9636    SO:0001583|missense_variant      1   \n",
       "4    0.00339  ...  ISG15:9636  SO:0001819|synonymous_variant      1   \n",
       "\n",
       "           RS                             CLNVI CLNSIGCONF CLNDISDBINCL  \\\n",
       "0   143888043                         Not_Given  Not_Given    Not_Given   \n",
       "1   150861311                         Not_Given  Not_Given    Not_Given   \n",
       "2   786201005  OMIM_Allelic_Variant:147571.0003  Not_Given    Not_Given   \n",
       "3  1553169766                         Not_Given  Not_Given    Not_Given   \n",
       "4    61766284                         Not_Given  Not_Given    Not_Given   \n",
       "\n",
       "   CLNDNINCL CLNSIGINCL    DBVARID  \n",
       "0  Not_Given  Not_Given  Not_Given  \n",
       "1  Not_Given  Not_Given  Not_Given  \n",
       "2  Not_Given  Not_Given  Not_Given  \n",
       "3  Not_Given  Not_Given  Not_Given  \n",
       "4  Not_Given  Not_Given  Not_Given  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_l = pd.concat([l, info_cols], axis=1)\n",
    "new_l = new_l.replace('.', 'Not_Given')\n",
    "new_l = new_l.fillna('Not_Given')\n",
    "new_l.drop('INFO', axis=1, inplace=True)\n",
    "new_l.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "The columns requested for presentation:\n",
    "\n",
    "1) Gene name\n",
    " `GENEINFO`\n",
    "2) Mutation ID number\n",
    "`ID`\n",
    "3) Mutation Position\n",
    "`CHROM & POS`\n",
    "4) Mutation value\n",
    "`REF & ALT` \n",
    "5) Clinical significance\n",
    "`CLNSIG`\n",
    "6) Disease that is implicated\n",
    "`CLNDN`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>GENEINFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>542074</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>542075</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS      ID REF ALT                  CLNSIG  \\\n",
       "0     1  1014O42  475283   G   A                  Benign   \n",
       "1     1  1O14122  542074   C   T  Uncertain_significance   \n",
       "2     1  1014143  183381   C   T              Pathogenic   \n",
       "3     1  1014179  542075   C   T  Uncertain_significance   \n",
       "4     1  1014217  475278   C   T                  Benign   \n",
       "\n",
       "                                               CLNDN    GENEINFO  \n",
       "0  Immunodeficiency_38_with_basal_ganglia_calcifi...  ISG15:9636  \n",
       "1  Immunodeficiency_38_with_basal_ganglia_calcifi...  ISG15:9636  \n",
       "2  Immunodeficiency_38_with_basal_ganglia_calcifi...  ISG15:9636  \n",
       "3  Immunodeficiency_38_with_basal_ganglia_calcifi...  ISG15:9636  \n",
       "4  Immunodeficiency_38_with_basal_ganglia_calcifi...  ISG15:9636  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_l=new_l[['CHROM', 'POS', 'ID', 'REF', 'ALT','CLNSIG','CLNDN','GENEINFO']]\n",
    "new_l.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "not_specified                                                                                   14400\n",
       "Not_Given                                                                                       12670\n",
       "Hereditary_cancer-predisposing_syndrome                                                          2951\n",
       "Limb-girdle_muscular_dystrophy,_type_2J|Dilated_cardiomyopathy_1G                                2588\n",
       "Lynch_syndrome                                                                                   1717\n",
       "not_specified|not_provided                                                                       1190\n",
       "Hereditary_nonpolyposis_colon_cancer                                                             1095\n",
       "Alstrom_syndrome                                                                                  846\n",
       "Nemaline_myopathy_2                                                                               769\n",
       "Cardiovascular_phenotype                                                                          742\n",
       "Limb-girdle_muscular_dystrophy,_type_2J|Dilated_cardiomyopathy_1G|not_specified                   673\n",
       "Brugada_syndrome                                                                                  661\n",
       "Neuroblastoma_3                                                                                   596\n",
       "Hereditary_nonpolyposis_colon_cancer|Hereditary_cancer-predisposing_syndrome                      585\n",
       "Ehlers-Danlos_syndrome,_type_4                                                                    562\n",
       "Limb-girdle_muscular_dystrophy,_type_2J|Dilated_cardiomyopathy_1G|not_specified|not_provided      496\n",
       "Primary_pulmonary_hypertension                                                                    441\n",
       "Catecholaminergic_polymorphic_ventricular_tachycardia                                             435\n",
       "Renal_hamartomas_nephroblastomatosis_and_fetal_gigantism                                          397\n",
       "Inborn_genetic_diseases                                                                           382\n",
       "Name: CLNDN, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CLNDN contains a list of disease names within the series\n",
    "new_l['CLNDN'].value_counts()[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Disease name column could be useful but i need to understand more about the context its being used in this dataset, since i'm not sure if it indeed suggests the person has this illness, possibly the Clinical significance colummn could provide more information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TTN:7273|TTN-AS1:100506866           6294\n",
       "TTN:7273                             4817\n",
       "MSH6:2956                            3507\n",
       "MSH2:4436                            2913\n",
       "MLH1:4292                            2337\n",
       "RYR2:6262                            1603\n",
       "BARD1:580                            1389\n",
       "NEB:4703                             1366\n",
       "SCN5A:6331                           1304\n",
       "ALMS1:7840                           1258\n",
       "USH2A:7399                           1195\n",
       "COL6A3:1293                          1070\n",
       "DYSF:8291                            1007\n",
       "MUTYH:4595                            955\n",
       "COL3A1:1281                           947\n",
       "ABCA4:24                              905\n",
       "LMNA:4000                             805\n",
       "ALK:238                               800\n",
       "SCN1A:6323|LOC102724058:102724058     724\n",
       "SCN1A:6323                            683\n",
       "Name: GENEINFO, dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GENEINFO contains a list of dictionaries within the series\n",
    "new_l['GENEINFO'].value_counts()[:20]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Gene names do not appear to be useful since I do not understand the signifance or risk level of specific genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Uncertain_significance                                                                      47980\n",
       "Likely_benign                                                                               17885\n",
       "Pathogenic                                                                                  12313\n",
       "Likely_pathogenic                                                                            6269\n",
       "Benign                                                                                       6138\n",
       "Conflicting_interpretations_of_pathogenicity                                                 5404\n",
       "Benign/Likely_benign                                                                         3338\n",
       "Not_Given                                                                                    1797\n",
       "Pathogenic/Likely_pathogenic                                                                  854\n",
       "risk_factor                                                                                    98\n",
       "association                                                                                    70\n",
       "drug_response                                                                                  41\n",
       "Affects                                                                                        31\n",
       "Conflicting_interpretations_of_pathogenicity,_risk_factor                                      14\n",
       "other                                                                                          12\n",
       "Pathogenic,_risk_factor                                                                        11\n",
       "Conflicting_interpretations_of_pathogenicity,_other                                             8\n",
       "Pathogenic/Likely_pathogenic,_other                                                             8\n",
       "Benign/Likely_benign,_risk_factor                                                               7\n",
       "protective                                                                                      6\n",
       "Pathogenic/Likely_pathogenic,_risk_factor                                                       6\n",
       "Likely_pathogenic,_risk_factor                                                                  5\n",
       "protective,_risk_factor                                                                         4\n",
       "Pathogenic,_other                                                                               2\n",
       "Pathogenic,_Affects                                                                             2\n",
       "Uncertain_significance,_risk_factor                                                             1\n",
       "Benign/Likely_benign,_protective                                                                1\n",
       "Affects,_risk_factor                                                                            1\n",
       "Conflicting_interpretations_of_pathogenicity,_protective                                        1\n",
       "Likely_pathogenic,_other                                                                        1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects,_other                                    1\n",
       "Benign,_drug_response                                                                           1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects,_association,_drug_response,_other        1\n",
       "Pathogenic,_protective                                                                          1\n",
       "Benign,_other                                                                                   1\n",
       "Likely_pathogenic,_association                                                                  1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects                                           1\n",
       "Pathogenic,_association,_protective                                                             1\n",
       "Likely_benign,_other                                                                            1\n",
       "Benign/Likely_benign,_other                                                                     1\n",
       "Conflicting_interpretations_of_pathogenicity,_other,_risk_factor                                1\n",
       "Conflicting_interpretations_of_pathogenicity,_association                                       1\n",
       "Benign/Likely_benign,_association                                                               1\n",
       "Name: CLNSIG, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_l['CLNSIG'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "The clinical significance column appear provide the neccessary level of detail to make an assumption about the risk of gene combindations.\n",
    "\n",
    "I observe that many columns contain the words pathogenic, i assume that the records containing this language will most often contain a dangerous combination."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Pathogenic', 'Conflicting_interpretations_of_pathogenicity',\n",
       "       'Likely_pathogenic', 'Pathogenic/Likely_pathogenic',\n",
       "       'Pathogenic,_risk_factor',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_other,_risk_factor',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_risk_factor',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_other',\n",
       "       'Pathogenic/Likely_pathogenic,_risk_factor',\n",
       "       'Likely_pathogenic,_risk_factor',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_association',\n",
       "       'Pathogenic/Likely_pathogenic,_other', 'Pathogenic,_other',\n",
       "       'Pathogenic,_association,_protective',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects',\n",
       "       'Likely_pathogenic,_association', 'Pathogenic,_protective',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects,_association,_drug_response,_other',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects,_other',\n",
       "       'Pathogenic,_Affects', 'Likely_pathogenic,_other',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_protective'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_l['CLNSIG'][new_l['CLNSIG'].str.contains('pathogenic',case=False)].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We take the data that contains the word pathogenic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "75.65895563960477"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_signif = new_l[new_l['CLNSIG'].str.contains('pathogenic',case=False)]\n",
    "(len(new_l)-len(df_signif))/len(new_l) * 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We've removed nearly $25%$% of records by using pathogenic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pathogenic                                                                                  12313\n",
       "Likely_pathogenic                                                                            6269\n",
       "Conflicting_interpretations_of_pathogenicity                                                 5404\n",
       "Pathogenic/Likely_pathogenic                                                                  854\n",
       "Conflicting_interpretations_of_pathogenicity,_risk_factor                                      14\n",
       "Pathogenic,_risk_factor                                                                        11\n",
       "Conflicting_interpretations_of_pathogenicity,_other                                             8\n",
       "Pathogenic/Likely_pathogenic,_other                                                             8\n",
       "Pathogenic/Likely_pathogenic,_risk_factor                                                       6\n",
       "Likely_pathogenic,_risk_factor                                                                  5\n",
       "Pathogenic,_other                                                                               2\n",
       "Pathogenic,_Affects                                                                             2\n",
       "Conflicting_interpretations_of_pathogenicity,_association                                       1\n",
       "Conflicting_interpretations_of_pathogenicity,_other,_risk_factor                                1\n",
       "Pathogenic,_association,_protective                                                             1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects                                           1\n",
       "Likely_pathogenic,_association                                                                  1\n",
       "Pathogenic,_protective                                                                          1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects,_association,_drug_response,_other        1\n",
       "Conflicting_interpretations_of_pathogenicity,_Affects,_other                                    1\n",
       "Likely_pathogenic,_other                                                                        1\n",
       "Conflicting_interpretations_of_pathogenicity,_protective                                        1\n",
       "Name: CLNSIG, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_signif['CLNSIG'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Conflicting interpretations` does not appear to be a dangerous mutation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Conflicting_interpretations_of_pathogenicity',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_other,_risk_factor',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_risk_factor',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_other',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_association',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects,_association,_drug_response,_other',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_Affects,_other',\n",
       "       'Conflicting_interpretations_of_pathogenicity,_protective'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_signif['CLNSIG'][df_signif['CLNSIG'].str.contains('Conflicting_interpretations')].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll omit `Conflicting interpretations` from the results set since it does not appear to signify a definitive disease risk level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "signif_pathogens=df_signif[~df_signif['CLNSIG'].str.contains('Conflicting_interpretations')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19474, 8)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "signif_pathogens.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>GENEINFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>1014316</td>\n",
       "      <td>161455</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>1014359</td>\n",
       "      <td>161454</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "      <td>ISG15:9636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>1022225</td>\n",
       "      <td>243036</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "      <td>AGRN:375790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>1022313</td>\n",
       "      <td>243037</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "      <td>AGRN:375790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>852</th>\n",
       "      <td>1</td>\n",
       "      <td>2406528</td>\n",
       "      <td>197886</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>PEX10:5192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>1</td>\n",
       "      <td>2406561</td>\n",
       "      <td>282334</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Not_Given</td>\n",
       "      <td>PEX10:5192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>1</td>\n",
       "      <td>2406580</td>\n",
       "      <td>555297</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "      <td>PEX10:5192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>1</td>\n",
       "      <td>2406580</td>\n",
       "      <td>296273</td>\n",
       "      <td>CAG</td>\n",
       "      <td>C</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Zellweger_syndrome|Peroxisome_biogenesis_disor...</td>\n",
       "      <td>PEX10:5192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>1</td>\n",
       "      <td>2406599</td>\n",
       "      <td>553074</td>\n",
       "      <td>GCT</td>\n",
       "      <td>G</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>Peroxisome_biogenesis_disorder_6A|Peroxisome_b...</td>\n",
       "      <td>PEX10:5192</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHROM      POS      ID  REF ALT             CLNSIG  \\\n",
       "2       1  1014143  183381    C   T         Pathogenic   \n",
       "8       1  1014316  161455    C  CG         Pathogenic   \n",
       "9       1  1014359  161454    G   T         Pathogenic   \n",
       "24      1  1022225  243036    G   A         Pathogenic   \n",
       "26      1  1022313  243037    A   T         Pathogenic   \n",
       "..    ...      ...     ...  ...  ..                ...   \n",
       "852     1  2406528  197886    G  GC         Pathogenic   \n",
       "857     1  2406561  282334    C   A         Pathogenic   \n",
       "862     1  2406580  555297   CA   C  Likely_pathogenic   \n",
       "863     1  2406580  296273  CAG   C         Pathogenic   \n",
       "866     1  2406599  553074  GCT   G  Likely_pathogenic   \n",
       "\n",
       "                                                 CLNDN     GENEINFO  \n",
       "2    Immunodeficiency_38_with_basal_ganglia_calcifi...   ISG15:9636  \n",
       "8    Immunodeficiency_38_with_basal_ganglia_calcifi...   ISG15:9636  \n",
       "9    Immunodeficiency_38_with_basal_ganglia_calcifi...   ISG15:9636  \n",
       "24                      Congenital_myasthenic_syndrome  AGRN:375790  \n",
       "26                      Congenital_myasthenic_syndrome  AGRN:375790  \n",
       "..                                                 ...          ...  \n",
       "852                                          Not_Given   PEX10:5192  \n",
       "857                                          Not_Given   PEX10:5192  \n",
       "862  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...   PEX10:5192  \n",
       "863  Zellweger_syndrome|Peroxisome_biogenesis_disor...   PEX10:5192  \n",
       "866  Peroxisome_biogenesis_disorder_6A|Peroxisome_b...   PEX10:5192  \n",
       "\n",
       "[100 rows x 8 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "signif_pathogens.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "4) Please Write your assumptions here:\n",
    "\n",
    "- Records with clinical significance containing the term \"pathogen\" that are not described as \"Conflicting interpretations\" suggest this mutation is potentially dangerous. The columns omitted from the final results do not contain relevant information to identify potentially dangerous mutations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "4) Findings / What would you present to your boss?\n",
    "\n",
    "- I've prepared a list of 19k pathogenic and likely pathogenic mutations of diseases, the first 100 of these obversations are listed in the attached file. Not all mutations and gene combindations have an indication of the disease they are likely to cause but they have been included since they are likely to be flagged as pathogenic. Further analysis might include grouping genes by likely disease to share information with various research groups looking to identify causes of various dangerous diseases."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
